vowpalwabbit/interactions.cc

```diff
@@ -4,123 +4,165 @@
+          const int nsc = *ns;
+          num_features_in_inter *= feature_spaces[nsc].size();
+          sum_feat_sq_in_inter *= feature_spaces[nsc].sum_feat_sq;
+          if (num_features_in_inter == 0) break;  // one of namespaces has no features - go to next interaction
         }
-      }
-    }
-  }
-}
+        else  // we are at beginning of a block made of same namespace (interaction is preliminary sorted)
+        {
+          // let's find out real length of this block
+          size_t order_of_inter = 2;  // alredy compared ns == ns+1
 
-// expand namespace interactions if contain wildcards
-// called from parse_args.cc
-// process all interactions in a vector
+          for (auto ns_end = ns + 2; ns_end < inter.end(); ++ns_end)
+            if (*ns == *ns_end) ++order_of_inter;
 
-std::vector<std::vector<namespace_index>> expand_interactions(
-    const std::vector<std::vector<namespace_index>>& vec, const size_t required_length, const std::string& err_msg)
-{
-  std::vector<std::vector<namespace_index>> res;
+          // namespace is same for whole block
+          const features& fs = feature_spaces[static_cast<int>(*ns)];
 
-  for (auto const& i : vec)
-  {
-    const size_t len = i.size();
-    if (required_length > 0 && len != required_length)
-    // got strict requirement of interaction length and it was failed.
-    {
-      THROW(err_msg);
-    }
-    else if (len < 2)
-      // regardles of required_length value this check is always performed
-      THROW("error, feature interactions must involve at least two namespaces" << err_msg);
+          // count number of features with value != 1.;
+          size_t cnt_ft_value_non_1 = 0;
 
-    std::vector<namespace_index> temp;
-    expand_namespaces_with_recursion(i, res, temp, 0);
-  }
-  return res;
-}
+          // in this block we shall calculate number of generated features and sum of their values
+          // keeping in mind rules applicable for simple combinations instead of permutations
 
-void expand_quadratics_wildcard_interactions(namespace_interactions& interactions)
-{
-  if (interactions.all_seen_namespaces_size == interactions.all_seen_namespaces.size())
-  {
-    // nothing new here
-    return;
-  }
+          // let's calculate sum of their squared value for whole block
 
-  interactions.all_seen_namespaces_size = interactions.all_seen_namespaces.size();
-  auto& set_interactions = interactions.all_seen_namespaces;
-  for (auto it = set_interactions.begin(); it != set_interactions.end(); ++it)
-  {
-    if (interactions.active_interactions.find({*it, *it}) == interactions.active_interactions.end())
-    {
-      interactions.interactions.push_back({*it, *it});
-      interactions.active_interactions.insert({*it, *it});
-    }
+          // ensure results as big as order_of_inter and empty.
+          for (size_t i = 0; i < results.size(); ++i) results[i] = 0.;
+          while (results.size() < order_of_inter) results.push_back(0.);
```

Comment:
nit:
`if(results.size() < order_of_inter) results.resize_but_with_stl_behavior(order_of_inter);`
The int constructor should initialize the new values to 0 (and if not, just swap the order of lines 92 and 93)`

edit: actually, result should just be the size of order_of_inter. Just ignore the if and always do the resize..